lib/sensors/sensors.hpp

```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <cstdint>
+#include <cstdio>
+#include <optional>
+
+namespace hyped::sensors {
+
+enum class WriteResult { kFailure, kSuccess };
+
+/**
+ * If a sensor is to be used with an I2C mux, it must inherit from this abstract class.
+ */
+class II2cSensor {
+ public:
+  /*
+   * This function carries out the initilization steps for a particular sensor.
+   */
+  virtual bool configure()               = 0;
+  virtual WriteResult write()            = 0;
+  virtual std::optional<uint16_t> read() = 0;
```

Comment:
Do we need both read and write access at all times? Surely most users will only require one of these.

Reply:
On second thought, the write function is redundant since thats what configure essentially does. Will remove that. We definitely need read though.

Reply:
Maybe call it `I2cReader` then? For consistency.

Reply:
But the classes deriving this will be like `ITemperature` or `IAccelerometer`  which won't just be readers?

Reply:
FWIW, we need a generic II2cSensor interface specifically because of the Multiplexer class that needs to be implemented soon. The Multiplexer class would need to accept II2c instances as a hardware mux could have any type of i2c sensor connected to it.

Reply:
> But the classes deriving this will be like `ITemperature` or `IAccelerometer` which won't just be readers?

I don't see anything wrong with this. Also, I don't think these will derive from this but rather wrap around it. I.e. they add some sort of code that interprets the `std::optional<uint16_t>` as a temperature or an acceleration and possibly does some setup.

If you really want to make these abstract (`ITemperature` vs `Temperature`) and it wasn't just a typo I don't think an accelerometer interface should have anything to do with I2C specifically.

> FWIW, we need a generic II2cSensor interface specifically because of the Multiplexer class that needs to be implemented soon. The Multiplexer class would need to accept II2c instances as a hardware mux could have any type of i2c sensor connected to it.

Let's chat tonight? Not sure I agree or understand.